### DISCUSSION TITLE
How to Manage 2 differents or more CityHub in Pybricks

### DISCUSSION THREAD

**Discussion by didiersm** [pybricks/support#1291] [General]
I have spent a lot of hours on this subject without found a solution.
In fact is it possible?
Tks

**Comment by laurensvalk** [comment]
Sure. But can you elaborate on what you would like to do?

You could make them communicate with the broadcast functionality.

If you mean programming them at the same time, you can simply open pybricks in another browser window.

**Reply by didiersm** [reply]
Thank you for your reply.
What I would like to achieve is:
  Train A leaves from point A, arrives at point B, Train B leaves from point C and arrives at point A...etc...to be simple.
In fact it is implementing the canton system.

**Reply by laurensvalk** [reply]
Perhaps you could place colored parkers on the tracks to mark those destinations. And place a color sensor facing down in the train base. 

If coordination between them is required, I think you could use the broadcast / observe functionality.

I've been thinking for a while to make a series of short video tutorials about trains, gradually increasing features and complexity. This could be one of them.

Any other applications that you'd like to see featured? Another more advanced video could cover distance keeping to avoid collisions.

**Reply by didiersm** [reply]
Thank you for the interest in my message.
I installed the color sensor and the colored blocks on the track. In shuttle mode it works very well with a single train.
It's the second one that's already giving me a problem..hihi...
I'm looking at the Brickrail app that some have used to automate.

**Reply by didiersm** [reply]
Hi Laurens

I try to manage 2 hubs as simpliest as it's possible
The Hubs communicate, it's OK but when the Hub1 (Train1) get the right color (RED) nothing happens on the other Hub (Tranway):

What is wrong with my code?
Could you help me please.


*****************************************************************************************
```python
#train1 
from pybricks.hubs import CityHub
from pybricks.pupdevices import DCMotor, Light, ColorDistanceSensor
from pybricks.parameters import Button, Color, Direction, Port, Stop
from pybricks.tools import wait, StopWatch
from urandom import randint

train1 = CityHub(broadcast_channel=1, observe_channels=[2,3])
trainmotor = DCMotor(Port.A)
trainsensor = ColorDistanceSensor(Port.B)

print("Program Started")

forward_speed = 40
check_color_interval_ms = 100

def check_for_color(color):
    while trainsensor.color() != color:
        wait(check_color_interval_ms)
	wait(100)
while True:
	color = trainsensor.color()
        train1.ble.broadcast(str(color))
        print(color)
```
********************************************************************************
```python
#tramway
from pybricks.hubs import CityHub
from pybricks.pupdevices import DCMotor, Light, ColorDistanceSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.tools import wait, StopWatch

tramwayHub = CityHub(observe_channels=[1], broadcast_channel=3)
tramwayMotor = DCMotor(Port.A)
tramwaysensor = ColorDistanceSensor(Port.B)
check_color_interval_ms = 20

def check_for_color(color):
    while True:
        # Receive broadcast from the other hub.

        data = tramwayHub.ble.observe(1)

        if data is None:
            tramwayHub.light.on(Color.RED)
        else:
            tramwayHub.light.on(Color.BLUE)
            rx_color = data[0]

            if str(color) == rx_color:
                return      
    wait(100)

print("Program tramway Started")
while True:
    print("Waiting tramway for RED")

    check_for_color(Color.RED)
    print("RED tramway detected")

    tramwayHub.light.on(Color.GREEN)
    print("Run Motor tramway forwards")
    tramwayMotor.dc(40)
    wait(1500)
    print("Stop tramway Motor")
    tramwayMotor.stop()
    wait(1500)
    print("Waiting tramway for not Red")
    check_for_color(Color.NONE)
```

**Reply by didiersm** [reply]
Sorry for the presentation, I don't how to put the code in a right way

**Reply by dlech** [reply]
> I don't how to put the code in a right way

I fixed it for you. You have to add a "code fence" with three backticks before and after the code like this:

````
```python
# code goes here
```
````

**Reply by dlech** [reply]
You either need to change this line in `#train1`



> ```python
> train1.ble.broadcast(str(color))
> ```


to 

```python
train1.ble.broadcast([str(color)])
```

or change these lines in `#tramway`


> ```python
>             rx_color = data[0]
> 
>             if str(color) == rx_color:
>                 return  
> ```

to 
```python
             if str(color) == data:
                return  
```

**Reply by BertLindeman** [reply]
Loose from the solid advice from David.

Just a naive glance at the code....

In your tramway code the wait(100) as it is now:
```python

def check_for_color(color):
    while True:
        # Receive broadcast from the other hub.

        data = tramwayHub.ble.observe(1)

        if data is None:
            tramwayHub.light.on(Color.RED)
        else:
            tramwayHub.light.on(Color.BLUE)
            rx_color = data[0]

            if str(color) == rx_color:
                return      
    wait(100)
```
Should maybe be indented 4 bytes extra? So it is within the whilw True.

Bert

**Reply by didiersm** [reply]
Thanks to all of you its' works  and thanks to the embeded code ;)

**Reply by didiersm** [reply]
I need still help
The motor goes on but never stop with a red brick!
Tks

```python
from pybricks.hubs import CityHub
from pybricks.pupdevices import DCMotor, Light, ColorDistanceSensor
from pybricks.parameters import Button, Color, Direction, Port, Stop
from pybricks.tools import wait, StopWatch

trainhub = CityHub(broadcast_channel=1, observe_channels=[2,3])
trainmotor = DCMotor(Port.A)
trainsensor = ColorDistanceSensor(Port.B)


def check_for_color(color):
    #print("Program Started")
    pass
while True:
    color = trainsensor.color()
    trainhub.ble.broadcast([str(color)])
    # broadcasts are sent every 100ms so no reason to update faster
    wait(100)

    while trainsensor.color() !="RED":
        trainmotor.dc(50)
    print("RED")
    check_for_color(Color.NONE)

```

**Reply by laurensvalk** [reply]
To debug something like this, it can help to print what the sensor measures:

```python
print(trainsensor.color())
```
You'll find that instead of ``"RED"``, you can use use ``Color.RED``.

See also this example: https://docs.pybricks.com/en/latest/pupdevices/colordistancesensor.html#waiting-for-a-color

We also have a brand new [block coding interface](https://pybricks.com/about/new-pybricks-blocks/) that may help. Just scroll upwards on the page linked above to see what all the blocks look like for the color sensor.
